keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,945 @@
++       GetPodVolumes(logger klogr.Logger, pod *v1.Pod) (boundClaims, unboundClaimsDelayBinding, unboundClaimsImmediate []*v1.PersistentVolumeClaim, err error)
+ 
+        // FindPodVolumes checks if all of a Pod's PVCs can be satisfied by the
+        // node and returns pod's volumes information.
+```
+
+##### Unit testing
+
+Each test case gets its own logger instance that adds log messages to the
+output of that test case. The log level can be configured with `-v` and
+`-vmodule` is supported.
+
+```
+diff --git a/pkg/scheduler/framework/plugins/volumebinding/binder_test.go b/pkg/scheduler/framework/plugins/volumebinding/binder_test.go
+index e354f18198c..debc08bc231 100644
+--- a/pkg/scheduler/framework/plugins/volumebinding/binder_test.go
++++ b/pkg/scheduler/framework/plugins/volumebinding/binder_test.go
+@@ -43,7 +43,7 @@ import (
+        "k8s.io/client-go/kubernetes/fake"
+        k8stesting "k8s.io/client-go/testing"
+        featuregatetesting "k8s.io/component-base/featuregate/testing"
+-       "k8s.io/klog/v2"
++       "k8s.io/klogr"
+        "k8s.io/kubernetes/pkg/controller"
+        pvtesting "k8s.io/kubernetes/pkg/controller/volume/persistentvolume/testing"
+        pvutil "k8s.io/kubernetes/pkg/controller/volume/persistentvolume/util"
+@@ -124,10 +124,6 @@ var (
+        zone1Labels = map[string]string{v1.LabelFailureDomainBetaZone: "us-east-1", v1.LabelFailureDomainBetaRegion: "us-east-1a"}
+ )
+ 
+-func init() {
+-       klog.InitFlags(nil)
+-}
+-
+...
+@@ -972,11 +969,12 @@ func TestFindPodVolumesWithoutProvisioning(t *testing.T) {
+        }
+ 
+        run := func(t *testing.T, scenario scenarioType, csiStorageCapacity bool, csiDriver *storagev1.CSIDriver) {
+-               ctx, cancel := context.WithCancel(context.Background())
++               logger, ctx := klogr.NewTestContext(t)
++               ctx, cancel := context.WithCancel(ctx)
+                defer cancel()
+
+diff --git a/pkg/scheduler/framework/plugins/volumebinding/main_test.go b/pkg/scheduler/framework/plugins/volumebinding/main_test.go
+new file mode 100644
+index 00000000000..00e882da70d
+--- /dev/null
++++ b/pkg/scheduler/framework/plugins/volumebinding/main_test.go
+@@ -0,0 +1,27 @@
+...
++import (
++       "testing"
++
++       "k8s.io/component-base/logs"
++)
++
++func TestMain(m *testing.M) {
++       logs.RunTests(m)
++}
+```
+
+
+##### Injecting common names and values, logger passed through existing ctx parameter
+
+```
+@@ -671,7 +692,10 @@ func (f *frameworkImpl) RunFilterPlugins(
+        nodeInfo *framework.NodeInfo,
+ ) framework.PluginToStatus {
+        statuses := make(framework.PluginToStatus)
++       logger := klogr.FromContext(ctx).WithName("Filter").WithValues("pod", klogr.KObj(pod), "node", klogr.KObj(nodeInfo.Node()))
+        for _, pl := range f.filterPlugins {
++               logger := logger.WithName(pl.Name())
```

Comment:
propose to use a diff name on the inside to avoid the potential `:=` bug

Reply:
I'm not sure about that proposal. Shadowing the outer variable is intentional: it makes it possible to cut-and-paste `logger.Info("some message")` around arbitrarily without having to remember which `logger` variable should be used in which part of the code.

Reply:
I wont argue more, but it seems more likely that I would notice (in review) you using the wrong logger more easily than you forgetting a `:`

Reply:
That depends. Suppose we have `loggerWithName := ...` here. Then in a future PR, someone adds a `logger.Info` call so far away from this line that none of the surrounding patch context uses `loggerWithName`. I'm pretty sure that would slip through the review.